### PR TITLE
Migrate bound values into bound literals

### PR DESCRIPTION
**Pull request by brusherru**
There is no separate issue, but it could be a part (or replacement) of #1164

### CODE REVIEW

File: packages/xod-project/src/extractBoundInputsToConstNodes.js [hunk 1]
```diff
@@ -113,12 +113,15 @@ const getMapOfExtractablePinPaths = def(
...
-            : CONST_NODETYPES[pinType]
-      )
+      R.map(([pinType, pinValue]) => {
+        if (pinType === PIN_TYPE.PULSE) {
+          return R.compose(
+            foldMaybe(null, R.prop(R.__, PULSE_CONST_NODETYPES)),
+            ensureLiteral
+          )(pinType, pinValue);
+        }
+        return CONST_NODETYPES[pinType];
```
Comment (nkrkv) [inline packages/xod-project/src/extractBoundInputsToConstNodes.js]: Hard to understand.

FP vs plain old JS turbulence detected. Looking at the two top functions of the composition I see you’re trying to (a) convert pinType + pinValue to paths which can fail, then (b) reject failures. Nulls/maybes mixed which leads to hard readability. What if:

```js
R.compose(
  R.map(catMaybies)
  R.map(stuffToRejectNothings)
  R.map(R.map((...) => pinType === PIN_TYPE.PULSE
    ? maybeProp(PULSE_CONST_NODETYPES, ensureLiteral(...))
    : maybeProp(CONST_NODETYPES, pinType)))
```

File: packages/xod-project/src/flatten.js [hunk 2]
```diff
@@ -732,7 +727,7 @@ const rekeyBoundValues = R.curry((boundValues, node) => {
     })(boundValues);
   }
 
-  return R.prop('boundValues', node);
+  return R.prop('boundLiterals', node);
```
Comment (nkrkv) [inline packages/xod-project/src/flatten.js]: Why not `getAllBoundValues`?

File: packages/xod-project/src/flatten.js [hunk 3]
```diff
@@ -783,12 +778,19 @@ export const extractPatches = R.curry(
...
+                'boundLiterals',
+                rekeyBoundValues(boundValues, node),
+                node
+              )
           ),
           R.converge(extractPatches(project, leafPaths), [
             R.compose(getPrefixedId(prefix), Node.getNodeId),
-            R.prop('boundValues'),
-            R.compose(getPatchByPath(project), Node.getNodeType),
+            R.prop('boundLiterals'),
```
Comment (nkrkv) [inline packages/xod-project/src/flatten.js]: Use API?

File: packages/xod-project/src/legacy.js [hunk 4]
```diff
@@ -33,12 +40,117 @@ export const ensureLiteral = def(
...
+  )(node);
+});
+
+// :: [Pin] -> Node -> Node
+const migrateNodeBoundValues = R.curry((node, pins) => {
+  const pinKeys = R.map(Pin.getPinKey, pins);
+  return R.compose(
+    // and then convert all other bound values
+    R.reduce(convertBoundValueToBoundLiteral, R.__, pins),
+    // omit bound values to unexisting pins
```
Comment (nkrkv) [inline packages/xod-project/src/legacy.js]: nonexistent

File: packages/xod-project/src/legacy.js [hunk 5]
```diff
@@ -33,12 +40,117 @@ export const ensureLiteral = def(
...
+    Patch.upsertNodes(R.__, patch),
+    R.map(node =>
+      R.compose(
+        // Remove `boundValues` completely if it's empty
+        R.when(
+          R.pipe(R.prop(LEGACY_BOUND_VALUES_PROP), R.isEmpty),
+          R.dissoc(LEGACY_BOUND_VALUES_PROP)
+        ),
+        R.ifElse(
+          Maybe.isNothing,
```
Comment (nkrkv) [inline packages/xod-project/src/legacy.js]: `R.ifElse(Maybe.isNothing` is a fold of a Maybe, isn’t it?

File: packages/xod-func-tools/src/monads.js [hunk 6]
```diff
@@ -190,3 +188,8 @@ export const leftIf = def(
   (condition, leftFn, val) =>
     condition(val) ? Either.of(val) : Either.Left(leftFn(val))
 );
+
+export const maybeProp = def(
```
Comment (evgenykochetkov) [inline packages/xod-func-tools/src/monads.js]: 👍🏻

**Pull request merged**